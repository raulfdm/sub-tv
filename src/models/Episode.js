class Episode {
  constructor(name, link) {
    this._name = name;
    this._link = `https://www.tv-subs.com${link}`;
  }

  get name() {
    return `Episode ${this._name}`;
  }

  get link() {
    return this._link;
  }
}

module.exports = Episode;
