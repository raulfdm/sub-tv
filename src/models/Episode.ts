export class EpisodeModel {
  private _link: string;
  constructor(private _name: string, link: string) {
    this._link = `https://www.tv-subs.com${link}`;
  }

  get name() {
    return `Episode ${this._name}`;
  }

  get link() {
    return this._link;
  }
}
