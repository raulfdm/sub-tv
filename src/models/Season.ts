// @ts-nocheck
export class Season {
  constructor(seasonNumber, link) {
    this._name = seasonNumber;
    this._link = `https://www.tv-subs.com${link}`;
  }

  get name() {
    return `Season ${this._name}`;
  }
  get link() {
    return this._link;
  }
}
