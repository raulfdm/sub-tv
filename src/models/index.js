export { default as EpisodeModel } from './Episode';
export { default as SeasonModel } from './Season';
export { default as SeriesModel } from './Series';
export { default as SubtitleListModel } from './SubtitleList';
export { default as SubtitleModel } from './Subtitle';
