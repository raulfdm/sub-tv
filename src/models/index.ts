export * from './Episode';
export * from './Season';
export * from './Series';
export * from './Subtitle';
export * from './SubtitleList';
