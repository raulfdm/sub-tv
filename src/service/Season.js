import axios from 'axios';
import { JSDOM } from 'jsdom';

import { SeasonModel } from '../models';

const SEASON_END_POINT = 'https://www.tv-subs.com/tv/';

export const fetchSeasonPage = chosenSeriesName => {
  const url = SEASON_END_POINT + chosenSeriesName;
  return axios.get(url).then(res => res.data);
};

/* Todo: test */
export const filterElements = $el => {
  const TV_REGEX = /\/tv\//;
  const NUMBER_REGEX = /\d/;
  const outerHTML = $el.outerHTML;
  const innerText = $el.textContent;

  return TV_REGEX.test(outerHTML) && NUMBER_REGEX.test(innerText);
};

export const createSeasonFromNodeElement = (seasonElements = []) => {
  return seasonElements.map(({ text, href }) => new SeasonModel(text, href));
};

export const handleHTML = markup => {
  const dom = JSDOM.fragment(markup);
  const listOfLinks = Array.from(dom.querySelectorAll('a'));
  const filterSeasons = listOfLinks.filter(filterElements);

  return filterSeasons.map($el => {
    return {
      href: $el.href,
      text: $el.textContent,
    };
  });
};

export default {
  fetch: chosenSeriesName => {
    return fetchSeasonPage(chosenSeriesName)
      .then(handleHTML)
      .then(createSeasonFromNodeElement);
  },
};
