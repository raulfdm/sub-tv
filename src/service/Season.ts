import axios from 'axios';
import { JSDOM } from 'jsdom';
import { SeasonModel } from '../models/Season';

function getSeasonHtmlElements(htmlText: string): Node[] {
  const dom = new JSDOM(htmlText);
  const elements = dom.window.document.querySelectorAll('[href*="season"]');

  return Array.from(elements);
}

function filterNonSeason(elementList: Node[]) {
  return elementList.filter(e => !/other/i.test(e.textContent || ''));
}

function convertHtmlToSeason(elementList: Node[]) {
  return elementList.map(el => {
    const name = el.textContent!;
    //@ts-ignore
    const link = el.href;
    return new SeasonModel(name, link);
  });
}

function fetch(serieName: string): Promise<SeasonModel[]> {
  const url = `https://www.tv-subs.com/tv/${serieName}`;

  return axios
    .get(url)
    .then(r => r.data)
    .then(getSeasonHtmlElements)
    .then(filterNonSeason)
    .then(convertHtmlToSeason);
}

export const SeasonService = {
  fetch,
};
