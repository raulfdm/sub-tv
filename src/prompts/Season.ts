import { inquirer, spinner } from '../instances';

import { SeasonService } from '../service/Season';
import { SeriesModel } from '../models';

export async function SeasonPrompt({ series }: { series: SeriesModel }) {
  spinner.start('Fetching available seasons');
  const seasons = await SeasonService.fetch(series.value);
  spinner.stop();

  const question = {
    choices: seasons.map(s => s.name),
    message: 'Choose the season',
    name: 'season',
    type: 'list',
    filter: function(answer: string) {
      const result = seasons.find(season => season.name === answer);
      return result;
    },
  };

  //@ts-ignore
  return inquirer.prompt(question);
}
