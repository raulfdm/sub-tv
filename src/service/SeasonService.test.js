/**
 * @jest-environment jsdom
 */

import mockedAxios from 'axios';
import { JSDOM } from 'jsdom';
import seasonPageContent from './__data__/seasonPage';
import { SeasonModel } from '../models';

import { fetchSeasonPage, handleHTML, createSeasonFromNodeElement } from './Season';

jest.mock('axios');

describe('fn: fetchSeasonPage', () => {
  const mockedHTML = `
  <div>
    <ul>
      <li>hue</li>
      < li>ahsuas</li>
      <li>br</li>
      <li>oi</li>
    </ul>
  </div>
  `;
  let result = undefined;

  beforeAll(async () => {
    mockedAxios.get.mockImplementationOnce(() => Promise.resolve({ data: mockedHTML }));
    const markup = await fetchSeasonPage('black-lightning');
    result = JSDOM.fragment(markup);
  });

  it('should match snapshot', async () => {
    expect(result.querySelector('div')).toMatchSnapshot(mockedHTML);
  });
});

describe('fn: handleHTML', () => {
  const result = handleHTML(seasonPageContent);

  it('should return an array', () => {
    expect(result).toBeInstanceOf(Array);
  });

  it('should array has 7 elements (pre defined at dom structure)', () => {
    expect(result).toHaveLength(7);
  });

  it('should all elements contain "text" and "href"', () => {
    result.forEach(el => {
      expect(el).toHaveProperty('text');
      expect(el.text).not.toBeFalsy();

      expect(el).toHaveProperty('href');
      expect(el.href).not.toBeFalsy();
    });
  });
});

describe('fn: createSeasonFromNodeElement', () => {
  const handledHtml = [
    [
      { href: '/tv/game-of-thrones/season-1/', text: '1' },
      { href: '/tv/game-of-thrones/season-2/', text: '2' },
      { href: '/tv/game-of-thrones/season-3/', text: '3' },
      { href: '/tv/game-of-thrones/season-4/', text: '4' },
      { href: '/tv/game-of-thrones/season-5/', text: '5' },
      { href: '/tv/game-of-thrones/season-6/', text: '6' },
      { href: '/tv/game-of-thrones/season-7/', text: '7' },
    ],
    [
      { href: '/tv/mr-robot/season-1/', text: '1' },
      { href: '/tv/mr-robot/season-2/', text: '2' },
      { href: '/tv/mr-robot/season-3/', text: '3' },
    ],
  ];

  handledHtml.forEach((seasonPlainObjList, index) => {
    describe(`case ${index + 1}`, () => {
      const result = createSeasonFromNodeElement(seasonPlainObjList);
      it('should return an array', () => {
        expect(result).toBeInstanceOf(Array);
      });

      it('should array has the same length sent', () => {
        expect(result).toHaveLength(seasonPlainObjList.length);
      });

      it('should all elements be instanceof SeasonModel', () => {
        result.forEach(el => {
          expect(el).toBeInstanceOf(SeasonModel);
        });
      });
    });
  });
});
