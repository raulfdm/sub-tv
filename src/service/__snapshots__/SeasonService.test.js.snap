// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fn: fetchSeasonPage should match snapshot: 
  <div>
    <ul>
      <li>hue</li>
      < li>ahsuas</li>
      <li>br</li>
      <li>oi</li>
    </ul>
  </div>
   1`] = `
<div>
  
    
  <ul>
    
      
    <li>
      hue
    </li>
    
      &lt; li&gt;ahsuas
      
    <li>
      br
    </li>
    
      
    <li>
      oi
    </li>
    
    
  </ul>
  
  
</div>
`;
